File: src/spectrl/codecs/quantized.py
"""Linear or log1p quantization followed by shuffled unsigned integer words.

Reconstruction goes through _deterministic.expm1 rather than the platform's,
so the same token decodes to the same bits in every implementation. See that
module for why the platform routine is not enough.
"""

import math

import numpy as np

from .._format import DEFAULT_INTENSITY_SCALE, DEFAULT_MZ_PPM
from ._deterministic import expm1 as deterministic_expm1
from ._deterministic import log1p as deterministic_log1p
from .shuffle import shuffle


def validate(params):
    if set(params) - {"scale", "width", "log", "delta"} or not {"scale", "width"} <= set(params):
        raise ValueError("quantized encoding requires scale and width")
    scale = params["scale"]
    if type(scale) not in (int, float) or not math.isfinite(scale) or scale <= 0:
        raise ValueError("quantization scale must be finite and positive")
    if type(params["width"]) is not int or params["width"] not in (1, 2, 4, 8):
        raise ValueError("quantized word width must be 1, 2, 4, or 8")
    if any(type(params[k]) is not bool for k in ("log", "delta") if k in params):
        raise ValueError("quantized log and delta parameters must be booleans")


def indices(data, params):
    a = np.asarray(data, dtype=np.float64)
    if not np.isfinite(a).all() or np.any(a < 0):
        raise ValueError("quantization requires finite nonnegative values")
    with np.errstate(over="ignore", invalid="ignore"):
        scaled = (deterministic_log1p(a) if params.get("log", False) else a) * params["scale"]
        floor = np.floor(scaled)
        q = floor + (scaled - floor >= 0.5)
    if not np.isfinite(q).all() or np.any(q > 2**53 - 1):
        raise ValueError("quantized index exceeds the safe integer range")
    return q.astype("<u8")


def parameters(data, scale, *, log=False, delta=False):
    params = {"scale": scale}
    if log:
        params["log"] = True
    if delta:
        params["delta"] = True
    q = indices(data, params)
    maximum = int(q.max(initial=0))
    params["width"] = next(w for w in (1, 2, 4, 8) if maximum < 2 ** (8 * w))
    return params


def intensity_parameters(data, scale=DEFAULT_INTENSITY_SCALE):
    """Choose the default log1p intensity grid, refined for values below 1.

    log1p is nearly linear below 1, so a fixed scale gives an absolute floor
    there and zeroes small peaks of normalized spectra. When the smallest
    positive value m is below 1, the scale grows to scale * (m + 1) / (2 * m),
    which bounds every positive value by the relative error the fixed scale
    already allows at 1.
    """
    source = np.asarray(data, dtype=np.float64)
    positive = source[source > 0]
    minimum = float(positive.min()) if len(positive) else 1.0
    if minimum < 1:
        refined = scale / 2 * (minimum + 1) / minimum
        if not math.isfinite(refined) or refined > 2**53 - 1:
            raise ValueError("intensity scale is outside the supported range")
        scale = max(scale, math.ceil(refined))
    return parameters(source, scale, log=True)


def ppm_parameters(data, ppm=DEFAULT_MZ_PPM):
    """Choose a log1p grid with a checked pointwise relative error bound."""
    if type(ppm) not in (int, float) or not math.isfinite(ppm) or ppm <= 0:
        raise ValueError("ppm must be finite and positive")
    source = np.asarray(data, dtype=np.float64)
    if not np.isfinite(source).all() or np.any(source < 0):
        raise ValueError("ppm quantization requires finite nonnegative values")
    positive = source[source > 0]
    minimum = float(positive.min()) if len(positive) else 1.0
    relative = ppm * 1e-6
    # Calibrate the log1p bound at the smallest positive value. Reserve a
    # numerical margin and verify the actual reconstruction before accepting.
    log_step = float(deterministic_log1p(relative * (1 - 1e-7) * (minimum / (minimum + 1))))
    if log_step <= 0 or not math.isfinite(log_step):
        raise ValueError("ppm scale is outside the supported range")
    scale = 0.5 / log_step
    if not math.isfinite(scale) or scale > 2**53 - 1:
        raise ValueError("ppm scale is outside the supported range")
    params = parameters(source, math.ceil(scale), log=True, delta=True)
    recovered = deterministic_expm1(indices(source, params).astype(np.float64) / params["scale"])
    if np.any(np.abs(source - recovered) > source * relative):
        raise ValueError("quantized reconstruction exceeds the ppm bound")
    return params


def encode(data, _type, params):
    q = indices(data, params)
    width = params["width"]
    if np.any(q > min(2**53 - 1, 2 ** (8 * width) - 1)):
        raise ValueError("quantized index exceeds its word width")
    words = q.astype(f"<u{width}")
    if params.get("delta", False):
        words[1:] = words[1:] - words[:-1]
    blob = shuffle(words.tobytes(), width)
    recovered = decode(blob, _type, len(q), params)
    source = np.asarray(data, dtype=np.float64)
    half_step = 0.5 / params["scale"]
    with np.errstate(over="ignore"):
        bound = (source + 1) * deterministic_expm1(half_step) if params.get("log", False) else half_step
    if np.any(np.abs(source - recovered) > bound):
        raise ValueError("quantized reconstruction exceeds the rounding bound")
    return blob


def decode(blob, _type, count, params):
    width = params["width"]
    if len(blob) != count * width:
        raise ValueError("quantized byte count mismatch")
    words = np.frombuffer(shuffle(blob, width, inverse=True), dtype=f"<u{width}")
    if params.get("delta", False):
        words = np.cumsum(words, dtype=words.dtype)
    if np.any(words > 2**53 - 1):
        raise ValueError("quantized index exceeds the safe integer range")
    with np.errstate(over="ignore", invalid="ignore"):
        result = words.astype(np.float64) / params["scale"]
        if params.get("log", False):
            result = deterministic_expm1(result)
    if not np.isfinite(result).all():
        raise ValueError("quantized reconstruction is not finite")
    return result

File: js/src/quantized.ts
/** A shared quantized-word representation for linear and log1p mappings. */
import { expm1 as deterministicExpm1, log1p as deterministicLog1p } from "./deterministic.js"
import { byteShuffle, byteUnshuffle, type NumArray } from "./codecs.js"
import type { Parameters } from "./pipeline.js"
import { deltaShuffleWords, deltaUnshuffleWords } from "./delta.js"
import { DEFAULT_INTENSITY_SCALE, DEFAULT_MZ_PPM } from "./format.js"

export function validateQuantized(p: Parameters): void {
  if (Object.keys(p).some(k => !["scale", "width", "log", "delta"].includes(k)) || !("scale" in p) || !("width" in p)) throw Error("quantized encoding requires scale and width")
  if (typeof p.scale !== "number" || !Number.isFinite(p.scale) || p.scale <= 0) throw Error("quantization scale must be finite and positive")
  if (![1, 2, 4, 8].includes(p.width as number)) throw Error("quantized word width must be 1, 2, 4, or 8")
  for (const key of ["log", "delta"]) if (key in p && typeof p[key] !== "boolean") throw Error("quantized log and delta parameters must be booleans")
}
function indices(a: NumArray, p: Parameters): Float64Array {
  return Float64Array.from(a, value => {
    if (!Number.isFinite(value) || value < 0) throw Error("quantization requires finite nonnegative values")
    const q = Math.round((p.log ? deterministicLog1p(value) : value) * (p.scale as number))
    if (!Number.isSafeInteger(q)) throw Error("quantized index exceeds the safe integer range")
    return q
  })
}
export function quantizedParameters(a: NumArray, scale: number, log = false, delta = false): Parameters {
  const p: Parameters = { scale, ...(log ? { log: true } : {}), ...(delta ? { delta: true } : {}) }
  let maximum = 0
  for (const q of indices(a, p)) maximum = Math.max(maximum, q)
  p.width = [1, 2, 4, 8].find(w => maximum < 2 ** (8 * w))!
  return p
}
// log1p is nearly linear below 1, so a fixed scale zeroes small peaks of
// normalized spectra. When the smallest positive value m is below 1, the scale
// grows to scale * (m + 1) / (2 * m), which bounds every positive value by the
// relative error the fixed scale already allows at 1.
export function intensityParameters(a: NumArray, scale = DEFAULT_INTENSITY_SCALE): Parameters {
  let minimum = Infinity
  for (const value of a) if (value > 0) minimum = Math.min(minimum, value)
  if (minimum < 1) {
    const refined = scale / 2 * (minimum + 1) / minimum
    if (!Number.isFinite(refined) || refined > 2 ** 53 - 1) throw Error("intensity scale is outside the supported range")
    scale = Math.max(scale, Math.ceil(refined))
  }
  return quantizedParameters(a, scale, true)
}
export function ppmParameters(a: NumArray, ppm = DEFAULT_MZ_PPM): Parameters {
  if (typeof ppm !== "number" || !Number.isFinite(ppm) || ppm <= 0) throw Error("ppm must be finite and positive")
  let minimum = Infinity
  for (const value of a) {
    if (!Number.isFinite(value) || value < 0) throw Error("ppm quantization requires finite nonnegative values")
    if (value > 0) minimum = Math.min(minimum, value)
  }
  if (minimum === Infinity) minimum = 1
  const relative = ppm * 1e-6
  // Calibrate at the smallest positive value and reserve a numerical margin.
  const logStep = deterministicLog1p(relative * (1 - 1e-7) * (minimum / (minimum + 1)))
  const scale = Math.ceil(0.5 / logStep)
  if (!(logStep > 0) || !Number.isSafeInteger(scale) || scale <= 0) throw Error("ppm scale is outside the supported range")
  const p = quantizedParameters(a, scale, true, true)
  const q = indices(a, p)
  for (const [i, value] of a.entries()) {
    const recovered = deterministicExpm1(q[i]! / scale)
    if (Math.abs(value - recovered) > value * relative) throw Error("quantized reconstruction exceeds the ppm bound")
  }
  return p
}
export function encodeQuantized(a: NumArray, _type: number, p: Parameters): Uint8Array {
  const q = indices(a, p)
  const width = p.width as number
  const raw = new Uint8Array(q.length * width)
  const view = new DataView(raw.buffer)
  for (const [i, value] of q.entries()) {
    if (value >= 2 ** (width * 8)) throw Error("quantized index exceeds its word width")
    if (width === 1) view.setUint8(i, value)
    else if (width === 2) view.setUint16(i * 2, value, true)
    else if (width === 4) view.setUint32(i * 4, value, true)
    else view.setBigUint64(i * 8, BigInt(value), true)
  }
  const blob = p.delta ? deltaShuffleWords(raw, width) : byteShuffle(raw, width)
  const recovered = decodeQuantized(blob, _type, q.length, p)
  const halfStep = 0.5 / (p.scale as number)
  for (const [i, value] of a.entries()) {
    const bound = p.log ? (value + 1) * deterministicExpm1(halfStep) : halfStep
    if (Math.abs(value - recovered[i]!) > bound) throw Error("quantized reconstruction exceeds the rounding bound")
  }
  return blob
}
export function decodeQuantized(blob: Uint8Array, _type: number, count: number, p: Parameters): Float64Array {
  const width = p.width as number
  if (blob.length !== count * width) throw Error("quantized byte count mismatch")
  const raw = p.delta ? deltaUnshuffleWords(blob, width) : byteUnshuffle(blob, width)
  const view = new DataView(raw.buffer, raw.byteOffset, raw.byteLength)
  return Float64Array.from({ length: count }, (_, i) => {
    const q = width === 1 ? view.getUint8(i) : width === 2 ? view.getUint16(i * 2, true)
      : width === 4 ? view.getUint32(i * 4, true) : Number(view.getBigUint64(i * 8, true))
    if (!Number.isSafeInteger(q)) throw Error("quantized index exceeds the safe integer range")
    const scaled = q / (p.scale as number)
    // The shared expm1, not the platform's, so the same token decodes to the
    // same bits everywhere. See src/deterministic.ts.
    const value = p.log ? deterministicExpm1(scaled) : scaled
    if (!Number.isFinite(value)) throw Error("quantized reconstruction is not finite")
    return value
  })
}
